File: DevEnvironment/DataStructures/src/test/java/datastructures/doublylinkedlist/core/DoublyLinkedListTestData.java
package datastructures.doublylinkedlist.core;

import base.core.ArrayLists;
import base.core.Triple;
import base.interfaces.ITriple;
import datastructures.doublylinkedlist.interfaces.IDoublyLinkedListNode;
import datastructures.doublylinkedlist.interfaces.IDoublyLinkedListTestData;
import datastructures.doublylinkedlist.interfaces.IDoublyLinkedListData;
import java.util.ArrayList;
import java.util.List;
import testbase.core.CollectionValues;

/**
 * The DoublyLinkedListTestData class implements data of tests of a doubly linked list.
 */
public final class DoublyLinkedListTestData implements IDoublyLinkedListTestData {
    /**
     * The DoublyLinkedListTestData constructor.
     */
    public DoublyLinkedListTestData() {
    }

    /**
     * Gets the data of doubly linked lists.
     */
    @Override
    public List<IDoublyLinkedListData<Integer>> getData() {
        List<IDoublyLinkedListData<Integer>> data = new ArrayList<>();

        data.add(this.getListData(0));
        data.add(this.getListData(50));
        data.add(this.getListData(100));

        return data;
    }

    /**
     * Gets the updation data of a doubly linked list.
     */
    @Override
    public List<ITriple<String, Integer, List<Integer>>> getUpdationData() {
        List<ITriple<String, Integer, List<Integer>>> data = new ArrayList<>();

        data.add(Triple.make("addToFront", 1, ArrayLists.make(1)));
        data.add(Triple.make("addToFront", 2, ArrayLists.make(2, 1)));
        data.add(Triple.make("addToFront", 3, ArrayLists.make(3, 2, 1)));
        data.add(Triple.make("removeFromFront", null, ArrayLists.make(2, 1)));
        data.add(Triple.make("removeFromFront", null, ArrayLists.make(1)));
        data.add(Triple.make("addToBack", 2, ArrayLists.make(1, 2)));
        data.add(Triple.make("addToBack", 3, ArrayLists.make(1, 2, 3)));
        data.add(Triple.make("addToBack", 4, ArrayLists.make(1, 2, 3, 4)));
        data.add(Triple.make("addToBack", 5, ArrayLists.make(1, 2, 3, 4, 5)));
        data.add(Triple.make("removeFromBack", null, ArrayLists.make(1, 2, 3, 4)));
        data.add(Triple.make("removeFromBack", null, ArrayLists.make(1, 2, 3)));

        return data;
    }

    /**
     * Gets the data of list.
     */
    private IDoublyLinkedListData<Integer> getListData(int size) {
        List<Integer> values = CollectionValues.createListValues(size);
        List<IDoublyLinkedListNode<Integer>> nodes = createListNodes(values);

        return new DoublyLinkedListData<>(
            values,
            nodes);
    }

    /**
     * Creates nodes of a doubly linked list.
     */
    private static <TValue extends Comparable<TValue>> List<IDoublyLinkedListNode<TValue>> createListNodes(List<TValue> values) {
        List<IDoublyLinkedListNode<TValue>> result = new ArrayList<>(values.size());

        for (TValue value : values) {
            result.add(DoublyLinkedListNode.make(value));
        }

        return result;
    }
}

File: SystemCore/DataStructuresTest/doublylinkedlist/core/DoublyLinkedListTestData.cpp
#include "DoublyLinkedListTestData.h"
#include "DoublyLinkedListData.h"

using namespace test::datastructures;
using namespace test::datastructures::doublylinkedlist;

/**
 * The DoublyLinkedListTestData constructor.
 */
DoublyLinkedListTestData::DoublyLinkedListTestData()
{
}

/**
 * The DoublyLinkedListTestData destructor.
 */
DoublyLinkedListTestData::~DoublyLinkedListTestData()
{
}

/**
 * Gets the data of doubly linked lists.
 */
std::vector<DoublyLinkedListDataSharedPtr<int>> DoublyLinkedListTestData::GetData() const
{
    std::vector<DoublyLinkedListDataSharedPtr<int>> data;

    data.push_back(GetListData(0));
    data.push_back(GetListData(50));
    data.push_back(GetListData(100));

    return data;
}

/**
 * Gets the updation data of a doubly linked list.
 */
std::vector<std::tuple<std::wstring, int, std::vector<int>>> DoublyLinkedListTestData::GetUpdationData() const
{
    std::vector<std::tuple<std::wstring, int, std::vector<int>>> data;

    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"AddToFront", 1, {1}));
    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"AddToFront", 2, {2, 1}));
    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"AddToFront", 3, {3, 2, 1}));
    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"RemoveFromFront", -1, {2, 1}));
    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"RemoveFromFront", -1, {1}));
    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"AddToBack", 2, {1, 2}));
    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"AddToBack", 3, {1, 2, 3}));
    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"AddToBack", 4, {1, 2, 3, 4}));
    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"AddToBack", 5, {1, 2, 3, 4, 5}));
    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"RemoveFromBack", -1, {1, 2, 3, 4}));
    data.push_back(std::make_tuple<std::wstring, int, std::vector<int>>(L"RemoveFromBack", -1, {1, 2, 3}));

    return data;
}

/**
 * Gets the data of a list.
 */
DoublyLinkedListDataSharedPtr<int> DoublyLinkedListTestData::GetListData(size_t size) const
{
    std::vector<int> values = CreateListValues(size);
    std::vector<IDoublyLinkedListNodeSharedPtr<int>> nodes = CreateListNodes(values);

    DoublyLinkedListDataSharedPtr<int> data(new DoublyLinkedListData<int>(
        values,
        nodes));

    return data;
}

/**
 * Creates values of a doubly linked list.
 */
std::vector<int> DoublyLinkedListTestData::CreateListValues(size_t size) const
{
    std::vector<int> result(size);

    for (size_t i = 0; i < size; ++i)
    {
        int value = i + 1;
        result[i] = value;
    }

    return result;
}

/**
 * Creates nodes of a doubly linked list.
 */
template <typename T>
std::vector<IDoublyLinkedListNodeSharedPtr<T>> DoublyLinkedListTestData::CreateListNodes(const std::vector<T>& values) const
{
    std::vector<IDoublyLinkedListNodeSharedPtr<T>> result(values.size());

    for (size_t i = 0; i < values.size(); ++i)
    {
        T value = values[i];
        result[i] = DoublyLinkedListNode<T>::Make(value);
    }

    return result;
}
